Alert with binding

randomeizer opened the pull request:
Created `alert(item:...)` method overloads for `View`. These allow providing an optional `Bindable<Item?>` value to the `alert(item: ...)` parameter, then providing other closures for the title, actions, and message, each passed an honest item, if it's not `nil`.

For example, a model:

```swift
@Observable
class FeatureModel {
  struct ConfirmDelete: Equatable {}

  var confirmDelete: ConfirmDelete?

  func deleteButtonTapped() {
    self.confirmDelete = .init()
  }
  
  func deleteConfirmButtonTapped() {
    // do deletion
    confirmDelete = nil
  }

  func deleteCancelButtonPressed() {
    confirmDelete = nil
  }
}
```

...and a View:

```swift
struct ContentView: View {
  @ObservedObject var model: FeatureModel

  var body: some View {
    Button("Delete Everything") {
      model.deleteButtonTapped()
    }
    List {
      // ...
    }
    .alert("Are you sure?", item: self.$model.confirmDelete) { _ in
      Button("Delete", role: .destructive) {
        model.deleteConfirmButtonTapped()
      }
      Button("Nevermind", role: .cancel) {
        model.deleteCancelButtonPressed()
      }
    } message: { _ in
      Text("Deleting this item cannot be undone.")
    }
  }
}
```

randomeizer commented:
Feedback wanted, particularly on:

# Parameter Order
Should `title` or `item` go first? I've split the difference, trying to follow established order in existing overloads. Basically:

1. If `title` is a closure, `item` goes first, then `title`, then the rest.
2. If `title` is a `String`/`LocalizableStringKey`, then that goes first, then `item`, then the rest.

This applies across the `alert` and `confirmationDialog` overloads.

# isPresent()

Currently, I've deprecated the old read-only `isPresent()` function in `SwiftUINavigation` for older versions, and added a new read/write `isPresent()` in `SwiftUINavigationCore`, to allow access from TCA.

Not sure if this is the right approach, of if we should just have a new, differently-named (maybe private?) func/var for use by `alert/confirmationDialog`.

Thoughts?

stephencelis commented on Sources/SwiftUINavigationCore/Alert.swift:
```diff
@@ -0,0 +1,386 @@
+    ///   - message: A view builder returning the message for the alert given the current alert state.
+    @available(iOS 15, macOS 12, tvOS 15, watchOS 8, *)
+    public func alert<Item, A: View, M: View>(
+      _ titleKey: LocalizedStringKey,
```
Let's not worry about static titles for now. If folks want they can either ignore the unwrapped item, or they can use one of Apple's messier vanilla APIs.

stephencelis commented on Sources/SwiftUINavigationCore/Binding.swift:
```diff
@@ -0,0 +1,23 @@
+    /// - Returns: A binding to a boolean. Returns `true` if non-`nil`, otherwise `false`.
+    public func isPresent<Wrapped>() -> Binding<Bool>
+      where Value == Wrapped? {
+      .init(
+        get: { self.wrappedValue != nil },
+        set: { isPresent, transaction in
+          guard !isPresent else { return }
+          self.transaction(transaction).wrappedValue = nil
+        }
+      )
```
Can you preserve the existing implementation that extends `Optional` with a private `_isPresent` property that is called via dynamic member lookup? `Binding.init(get:set:)` does not propagate the current binding's SwiftUI transaction properly so this could break animations and transitions in certain contexts.

randomeizer replied:
Yeah, wasn't sure what the best approach was here. Will work on that.

stephencelis commented on Sources/SwiftUINavigationCore/Documentation.docc/Articles/Dialogs.md:
```diff
@@ -0,0 +1,101 @@
+# Dialogs
```
We already have an `AlertsDialogs.md` that is a good place for docs related to presenting alerts and dialogs. Was this added by mistake? Also, feel free to omit new DocC articles for now. We can consider updating existing docs closer to a version release :smile:

randomeizer replied:
If preserving, what would be the best approach to make it consistent?

Given there isn't really any other documentation in `Core` currently I think I'll just remove the new articles.

stephencelis reviewed (commented):
Hey @randomeizer! Thanks for taking this on! A few comments below that should be easy to tackle, and then I think it'll be good to merge!

randomeizer commented:
@stephencelis I believe I've addressed your comments. Maybe double-check the Binding migration - I wasn't 100% certain that I extracted all the relevant elements of `Binding` into `Core`, but it does compile and my app that uses this code works as expected.

stephencelis commented on Sources/SwiftUINavigation/Internal/Deprecations.swift:
```diff
@@ -2,6 +2,39 @@
+
+  @available(iOS 15, macOS 12, tvOS 15, watchOS 8, *)
+  extension View {
+    @available(*, deprecated, renamed: "alert(_:title:actions:message:)")
```
Should this include `item`?

```suggestion
    @available(*, deprecated, renamed: "alert(item:title:actions:message:)")
```

stephencelis commented on Sources/SwiftUINavigation/Internal/Deprecations.swift:
```diff
@@ -2,6 +2,39 @@
+      alert(item: value, title: title, actions: actions, message: message)
+    }
+    
+    @available(*, deprecated, renamed: "confirmationDialog(_:textVisibility:title:actions:message:)")
```
```suggestion
    @available(*, deprecated, renamed: "confirmationDialog(item:textVisibility:title:actions:message:)")
```

stephencelis commented on Sources/SwiftUINavigationCore/Documentation.docc/SwiftUINavigationCore.md:
```diff
@@ -1,12 +0,0 @@
-# ``SwiftUINavigationCore``
```
Do you know why this file got deleted?

randomeizer replied:
Ah, I assumed that I had added all the documentation, my bad.
